core/commands/storage.go

```diff
@@ -112,61 +113,123 @@ Receive proofs as collateral evidence after selected nodes agree to store the fi
 		cmds.StringOption(hostSelectModeOptionName, "m", "Based on mode to select the host and upload automatically.").WithDefault("score"),
-		cmds.StringOption(hostSelectionOptionName, "l", "Use only these hosts in order on 'custom' mode. Use ',' as delimiter."),
+		cmds.StringOption(hostSelectionOptionName, "s", "Use only select these hosts in order on 'custom' mode. Use ',' as delimiter."),
 	},
 	RunTimeout: 5 * time.Minute, // TODO: handle large file uploads?
 	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
-		p, found := req.Options[uploadPriceOptionName].(int64)
-		if found && p < 0 {
+		// get node
+		n, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+		// get core api
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+		// check file hash
+		var (
+			chunckHashes []string
+			rootHash     string
+		)
+		lf, _ := req.Options[leafHashOptionName].(bool)
+		if lf == false {
+			if len(req.Arguments) != 1 {
+				return fmt.Errorf("need one and only one root hash.")
+			}
+			// get leaf hashes
+			rootHash = req.Arguments[0]
+			hashToCid, err := cidlib.Parse(rootHash)
+			if err != nil {
+				return err
+			}
+			rp, err := api.ResolvePath(req.Context, path.IpfsPath(hashToCid))
+			if err != nil {
+				return err
+			}
+			cids, err := api.Object().Links(req.Context, rp)
+			if err != nil {
+				return err
+			}
+			for _, cid := range cids {
+				chunckHashes = append(chunckHashes, cid.Cid.String())
+			}
+		} else {
+			rootHash = ""
+			chunckHashes = req.Arguments
+		}
+		// set price limit
+		price, found := req.Options[uploadPriceOptionName].(int64)
+		if found && price < 0 {
 			return fmt.Errorf("cannot input a negative price")
 		} else if !found {
 			// TODO: Select best price from top candidates
-			p = int64(1)
+			price = int64(1)
 		}
-
+		// get hosts/peers
+		var peers []string
 		mode, _ := req.Options[hostSelectModeOptionName].(string)
-		hosts, found := req.Options[hostSelectionOptionName].(string)
-		if mode == "custom" && !found {
-			return fmt.Errorf("custom mode needs input host lists")
+		if mode == "custom" {
+			// get host list as user specified
+			hosts, found := req.Options[hostSelectionOptionName].(string)
+			if !found {
+				return fmt.Errorf("custom mode needs input host lists")
+			}
+			peers = strings.Split(hosts, ",")
+		} else {
+			// get host list from storage
+			rds := n.Repo.Datastore()
+			qr, err := rds.Query(query.Query{
+				Prefix: hostStorePrefix + mode,
+				Orders: []query.Order{query.OrderByKey{}},
+			})
+			if err != nil {
+				return err
+			}
+			for r := range qr.Next() {
+				if r.Error != nil {
+					return r.Error
+				}
+				var ni info.Node
+				err := json.Unmarshal(r.Entry.Value, &ni)
+				if err != nil {
+					return err
+				}
+				peers = append(peers, ni.NodeID)
+				if len(peers) == len(chunckHashes) {
+					break
```

Comment:
on early exit should call `qr.Close()`

ref: https://github.com/ipfs/go-datastore/blob/master/query/query.go#L208